status-im/status-react
# Swipe gestures for Activity Center notifications with CTA

Fixes https://github.com/status-im/status-mobile/issues/15118

### Summary

This PR implements swipe actions for notifications with call to action (e.g. pending contact requests, unverified identity verifications, etc).

According to the Design team, the goal is to deliver a consistent experience to users, so whenever the user sees a notification with buttons, the same actions can be taken via the swipe buttons.

**Swipe buttons are using placeholder icons while the Design team works out which ones to use**.

Additionally, a bunch of fixes :lady_beetle: 

- [x] Fix: Membership tab not showing unread indicator.
- [x] Fix: outgoing pending contact requests were not being removed from the UI when cancelled.
- [x] Fix: dismissed membership notification not marked as read.
- [x] Fix: dismissed membership notification was displaying decline/accept buttons. Regression came from changes in status-go related to soft deletion of notifications.
- [x] Fix: incorrect check for the pending state of a contact request.
- [x] Fixed lots of bugs for identity verification notifications, as it was completely broken. Unfortunately, somebody made lots of changes without actually testing the flows.
- [x] Add basic error handling and log if accepting, declining or canceling contact requests fail.

### Demo

This demo shows an identity verification with swipe actions to reply or decline.

[identity-verification-swipe-to-reply.webm](https://user-images.githubusercontent.com/46027/223565755-b2ca3f68-12e2-4e1e-9e52-edd52cfcc971.webm)

### Out of scope

The old quo input is still in use in the identity verification notification. This will eventually be solved by issue https://github.com/status-im/status-mobile/issues/14364

### Steps to test

Notifications with one or more buttons (actions) are affected by this PR, because now the user can also swipe left/right to act on them. Nonetheless, everything should behave exactly the same.

- Membership notifications: private group chat. The following PR explains how to generate them https://github.com/status-im/status-mobile/pull/14785
- Contact requests, and community gated requests to join (Admin tab).
- Identity verifications. I believe the only way to test identity verification flows at the moment is to use the Desktop app, since initiating the challenge is not implemented in Mobile yet.
- Mentions and replies don't have new swipe buttons because they don't have call to action buttons throughout their lifecycle.

Please, keep in mind the identity verification feature you see on the video will soon improve to use the new quo2 input component. There's also a known bug that the keyboard is covering the bottom sheet.

If you want to test identity verification flows, here are the basic flows:

#### Identity verification flow 1

- `A` and `B` are mutual contacts.
- `A` sends a verification request to `B`.
- `A` should not see any notification yet.
- `B` should receive an identity verification notification. `B` can either decline or reply.
- `B` declines and the status `Declined` is shown instead of buttons.
- `B` can now either swipe to toggle read/unread or swipe delete the notification.
- `A` should not receive any notification after `A` declined.

#### Identity verification flow 2

- `A` and `B` are mutual contacts.
- `A` sends a verification request to `B`.
- `A` should not see any notification yet.
- `B` should receive an identity verification notification. `B` can either decline or reply.
- `B` press `Reply` and a bottom sheet is displayed with a text input.
- `B` sends the reply/answer message and the status `Replied` is shown instead of buttons.
- `B` can now either swipe to toggle read/unread or swipe to delete the notification.
- `A` should receive a notification with the reply from `B`.
- `A` can either mark the answer as untrustworthy or accept it (trust it) via the normal buttons, as well as via the swipe left/right buttons.
- If `A` accepts the answer, then the status `Confirmed` is shown instead of buttons. On the other hand, if `A` marks as untrustworthy, then the status `Untrustworthy` is shown instead of buttons.
- `B` should receive no further notifications due to `A`s actions.
- `A` can now either swipe to toggle read/unread or swipe delete the
  notification.

status: ready
